py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/logging_utils.py

```diff
@@ -0,0 +1,72 @@
+import logging
+import time
+import uuid
+
+from typing import Optional, Union
+
+MdValues = Union[str, int, float]
+
+
+class ProcessLogger:
+    """
+    Class to help with logging events that happen inside of a function.
+    """
+
+    def __init__(
+        self, process_name: str, **metadata: Optional[MdValues]
+    ) -> None:
+        """
+        create a process logger with a name and optional metadata. a start time
+        and uuid will be created for timing and unique identification
+        """
+        self.process_name = process_name
+        self.uuid = uuid.uuid4()
+        self.start_time = time.time()
+        self.metadata = metadata
+
+    def add_metadata(self, **metadata: Optional[MdValues]) -> None:
+        """add metadata to the process logger"""
+        for key, value in metadata.items():
+            self.metadata[key] = value
+
+    def log_start(self) -> None:
+        """log the start of a proccess"""
+        logging_string = f"start={self.process_name}, uuid={self.uuid}"
+
+        if self.metadata:
+            metadata_string = ", ".join(
+                [f"{key}={value}" for (key, value) in self.metadata.items()]
+            )
+            logging_string += f", {metadata_string}"
+        logging.info(logging_string)
+
+    def log_complete(self) -> None:
+        """log the completion of a proccess with duration"""
+        duration = self.start_time - time.time()
+        logging_string = (
+            f"complete={self.process_name}, "
+            f"uuid={self.uuid}, "
+            f"durration={duration:.2f}"
+        )
+
+        if self.metadata:
+            metadata_string = ", ".join(
+                [f"{key}={value}" for (key, value) in self.metadata.items()]
+            )
+            logging_string += f", {metadata_string}"
+        logging.info(logging_string)
+
+    def log_failure(self, exception: Exception) -> None:
```

Comment:
failure logs everything start does, plus an error type and a traceback (by using logging.exception)